Add rewrites to remove unnecessary `expm1` operations

ricardoV94 opened the pull request:
An alternative strategy would be to canonicalize all these `log*` and `exp*` Ops and let the `add_mul_canonizer` do it's thing with the `+1/-1`, after which a single `log(exp(x)) -> x` `exp(log(x)) -> switch(x>=0, x, np.nan)` might be able to cover all cases...

This would require at least, removing the `local_expm1` rewrite that is registered already during canonicalization, and perhaps delay some other rewrites that build (or build on top of) these `Ops`.

brandonwillard reviewed (changes requested):
`new_out.type != old_out.type` is too strict and possibly unnecessary.  See the updated [`TensorType` documentation](https://aesara.readthedocs.io/en/latest/extending/type.html) for more information.

In general, the only `Type` filtering a rewrite should need to do is the kind that determines whether or not the rewrite is applicable in the first place.  For rewrites like these, the shapes before and after rewriting shouldn't change, so such checks are a sign that the boundaries between operations/concerns in the rewrite process aren't clear&mdash;in that the concerns of `Type` conversion have leaked into the logic for rewrites.  

**We need to start letting these kinds of conversions happen downstream** (e.g. in `FunctionGraph.replace`), unless they're a fundamental part of a rewrite itself (e.g. when explicit `dtype` conversion is an inextricable part of the rewrite itself).

Put another way: a falsy `return` in a `local_optimizer` means that the rewrite **doesn't apply** to the node, but, when a rewrite _does_ apply&mdash;but the types can't be converted for some reason&mdash;that's a separate type system issue.

Theano confused these distinct domains, which led to a lot of leaky logic, and, in many cases, these conditions actually serve as `assert`s, because the inequality conditions should never be true unless there's an actual typing error.  Worse yet, those `return` statements are hiding the errors.

ricardoV94 commented:
Yeah, I forgot about the new changes, and added those from muscle memory. 

I checked the specific rewrites affected by this PR, and the only special case that should be needed is when the simplification returns the original input `x`, in which case there may have been a cast from integer to float. I added a commit that removes the preexisting type checks in these two rewrites.
